boards: arm: Add support for nucleo_g071rb board

FRASTM opened the pull request:
This patch adds support for the STM32 nucleo_g071rb board from STMicroelectronics.

This patch requires  zephyrproject-rtos/hal_stm32#3

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: interrupt_controller: Add STM32G0X interrupt support

Add interrupt support for STM32G0X SoC series.

Signed-off-by: Philippe Retornaz <philippe@shapescale.com>
Signed-off-by: Francois Ramu <francois.ramu@st.com>

# Conflicts:
#	drivers/interrupt_controller/exti_stm32.c

FRASTM commented:
recheck

FRASTM commented:
recheck

dbkinder reviewed (approved):
docs LGTM, thanks!

erwango commented on west.yml:
```diff
@@ -51,7 +51,7 @@ manifest:
       path: modules/hal/st
     - name: hal_stm32
-      revision: b454ad819f4f10cb82ce5eb1d7f709a680bc61c5
+      revision: pull/3/head
```
pull/13/head

erwango replied:
This ommit will  have to be removed

erwango commented on dts/bindings/flash_controller/st,stm32g0-flash-controller.yaml:
```diff
@@ -0,0 +1,15 @@
+---
```
Remove this line

erwango commented on dts/bindings/flash_controller/st,stm32g0-flash-controller.yaml:
```diff
@@ -0,0 +1,15 @@
+    compatible:
+      constraint: "st,stm32g0-flash-controller"
+
+...
```
Remove this line

erwango commented on soc/arm/st_stm32/stm32g0/Kconfig.soc:
```diff
@@ -0,0 +1,17 @@
+prompt "STM32G0x MCU Selection"
+depends on SOC_SERIES_STM32G0X
+
+config SOC_STM32G071RB
```
SOC_STM32G071XX

erwango commented on soc/arm/st_stm32/stm32g0/Kconfig.soc:
```diff
@@ -0,0 +1,17 @@
+depends on SOC_SERIES_STM32G0X
+
+config SOC_STM32G071RB
+	bool "STM32G071RB"
```
STM32G071XX

erwango commented on soc/arm/st_stm32/stm32g0/Kconfig.defconfig.stm32g071rb:
```diff
@@ -0,0 +1,22 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+if SOC_STM32G071RB
```
Use `SOC_STM32G071XX`
We don't need to introduce additional information (RB) here. This would prevent code factorization

erwango commented on soc/arm/st_stm32/stm32g0/Kconfig.defconfig.stm32g071rb:
```diff
@@ -0,0 +1,22 @@
+if GPIO_STM32
+
+endif # GPIO_STM32
+endif # SOC_STM32G071RB
```
SOC_STM32G071XX

erwango commented on soc/arm/st_stm32/stm32g0/dts_fixup.h:
```diff
@@ -0,0 +1,17 @@
+
+#define DT_NUM_IRQ_PRIO_BITS	        DT_ARM_V6M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
+
+#define DT_PWM_STM32_3_DEV_NAME		    DT_ST_STM32_PWM_40000400_PWM_LABEL
```
PWM defines should be introduced in PWM related commit

erwango replied:
Not sure you you move these in this commit

erwango commented on soc/arm/st_stm32/stm32g0/soc_registers.h:
```diff
@@ -0,0 +1,14 @@
+/*
```
Remove this file

erwango commented on soc/arm/st_stm32/stm32g0/flash_registers.h:
```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2019 Philippe Retornaz <philippe@shapescale.com>
```
Add this file when adding Flash support

erwango commented on drivers/flash/CMakeLists.txt:
```diff
@@ -26,6 +26,7 @@ if(CONFIG_CLOCK_CONTROL_STM32_CUBE)
   zephyr_sources_ifdef(CONFIG_SOC_SERIES_STM32F7X flash_stm32f7x.c)
```
Have you been able to test flash driver?
(using sample/drivers/flash_shell)

FRASTM replied:
Testing samples/drivers/flash_shell  --> PASSED

uart:~$ flash erase 0x1E000 0x1000
uart:~$ flash write 0x1E000 00 01 02 03 04 05 06 07
Reading back written bytes:
00 01 02 03 | 04 05 06 07
uart:~$ flash read 0x1E000 10
00 01 02 03 | 04 05 06 07
ff ff

erwango commented on drivers/clock_control/clock_stm32g0.c:
```diff
@@ -0,0 +1,53 @@
+{
+#if defined(CONFIG_EXTI_STM32) || defined(CONFIG_USB_DC_STM32)
+	/* Enable System Configuration Controller clock. */
+	LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_SYSCFG);
```
Is that valid? Looking to previous file, it seemed APB2 was not supported

erwango replied:
This comment still stands

erwango commented on dts/arm/st/g0/stm32g0.dtsi:
```diff
@@ -46,9 +46,33 @@
+			label = "STM32_CLK_RCC";
+		};
+
+		timers3: timers@40000400 {
```
Keep this node for PWM commit

erwango commented on soc/arm/st_stm32/stm32g0/dts_fixup.h:
```diff
@@ -14,4 +14,7 @@
 #define DT_TIM_STM32_3_CLOCK_BITS		DT_ST_STM32_TIMERS_40000400_CLOCK_BITS
 #define DT_TIM_STM32_3_CLOCK_BUS		DT_ST_STM32_TIMERS_40000400_CLOCK_BUS
 
+#define DT_FLASH_DEV_BASE_ADDRESS		DT_ST_STM32G0_FLASH_CONTROLLER_40022000_BASE_ADDRESS
```
Keep this one for flash commit

erwango commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -54,7 +54,7 @@ static void config_bus_clk_init(LL_UTILS_ClkInitTypeDef *clk_init)
 					CONFIG_CLOCK_STM32_APB1_PRESCALER);
 
-#ifndef CONFIG_SOC_SERIES_STM32F0X
+#if !defined (CONFIG_SOC_SERIES_STM32F0X) && !defined (CONFIG_SOC_SERIES_STM32G0X)
```
If APB2 not supported, you also need to update:
`config CLOCK_STM32_APB2_PRESCALER` in `drivers/clock_control/Kconfig.stm32`

erwango replied:
This comment still stands

erwango commented on drivers/pinmux/stm32/pinmux_stm32.c:
```diff
@@ -107,15 +107,20 @@ static int enable_port(u32_t port, struct device *clk)
 
 static int stm32_pin_configure(int pin, int func, int altf)
 {
+#ifdef CONFIG_GPIO
```
This change should be done and explained in a separate PR

erwango replied:
This comment still stands

erwango commented on dts/arm/st/g0/stm32g0.dtsi:
```diff
@@ -106,6 +106,24 @@
+			status = "disabled";
+			label = "UART_2";
+		};
+
```
We miss the lpuart node

erwango commented on soc/arm/st_stm32/stm32g0/dts_fixup.h:
```diff
@@ -55,13 +55,30 @@
+#define DT_UART_STM32_USART_2_CLOCK_BITS	DT_ST_STM32_USART_40004400_CLOCK_BITS
+#define DT_UART_STM32_USART_2_CLOCK_BUS	DT_ST_STM32_USART_40004400_CLOCK_BUS
+#define DT_UART_STM32_USART_2_HW_FLOW_CONTROL	DT_ST_STM32_USART_40004400_HW_FLOW_CONTROL
+
```
Miss the lpuart

erwango commented on soc/arm/st_stm32/stm32g0/soc.h:
```diff
@@ -44,6 +44,11 @@
 
+#ifdef CONFIG_SERIAL_HAS_DRIVER
+#include <stm32g0xx_ll_usart.h>
+#include <stm32g0xx_ll_lpuart.h>
```
Either remove this line, either add the LPUART node support

erwango commented on boards/arm/nucleo_g071rb/doc/index.rst:
```diff
@@ -0,0 +1,186 @@
++-----------+------------+-------------------------------------+
+| CLOCK     | on-chip    | reset and clock control             |
++-----------+------------+-------------------------------------+
+| I2C       | on-chip    | i2c controller                      |
```
Not available yet

erwango commented on boards/arm/nucleo_g071rb/doc/index.rst:
```diff
@@ -0,0 +1,186 @@
++-----------+------------+-------------------------------------+
+| I2C       | on-chip    | i2c controller                      |
++-----------+------------+-------------------------------------+
+| SPI       | on-chip    | spi controller                      |
```
Not available yet

erwango commented on boards/arm/nucleo_g071rb/support/openocd.cfg:
```diff
@@ -0,0 +1,23 @@
+
+set WORKAREASIZE 0x2000
+
+source [find target/stm32g0.cfg]
```
Is that working?

erwango replied:
Question stands

FRASTM replied:
uart:~$ flash erase 0x1E000 0x1000
uart:~$ flash write 0x1E000 00 01 02 03 04 05 06 07
Reading back written bytes:
00 01 02 03 | 04 05 06 07
uart:~$ flash read 0x1E000 10
00 01 02 03 | 04 05 06 07
ff ff

FRASTM committed:
drivers: pwm: Add STM32G0XX pwm support

Add pwm support for STM32G0X SoC series.

Signed-off-by: Philippe Retornaz <philippe@shapescale.com
Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
soc: arm: uart: Add STM32G0X UART support

Add UART support for STM32G0X SoC series.

Signed-off-by: Philippe Retornaz <philippe@shapescale.com>
Signed-off-by: Francois Ramu <francois.ramu@st.com>

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
recheck

erwango commented on boards/arm/nucleo_g071rb/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library()
```
Need to protect under if(CONFIG_PINMUX).
Cf other boards following recent rebase

FRASTM replied:
done

erwango commented:
@FRASTM , can you take into account upcoming https://github.com/zephyrproject-rtos/zephyr/pull/17122?

erwango commented on soc/arm/st_stm32/stm32g0/dts_fixup.h:
```diff
@@ -0,0 +1,15 @@
+
+#define DT_NUM_IRQ_PRIO_BITS	        DT_ARM_V6M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
+
+#define DT_TIM_STM32_3_BASE_ADDRESS		DT_ST_STM32_TIMERS_40000400_BASE_ADDRESS
```
Please remove these

FRASTM committed:
soc: arm: st_stm32: Add STM32G071 SoC series

This patch adds support for the STM32G071xx
 from STMicroelectronics.

Signed-off-by: Philippe Retornaz <philippe@shapescale.com>
Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: pinmux: Add STM32G0X pinmux support

Add pinmux support for STM32G0X SoC series.

Signed-off-by: Philippe Retornaz <philippe@shapescale.com>
Signed-off-by: Francois Ramu <francois.ramu@st.com>

# Conflicts:
#	drivers/pinmux/stm32/pinmux_stm32.h

FRASTM committed:
drivers: gpio: Add STM32G0X pinmux support

Add gpio support for STM32G0X SoC series.

Signed-off-by: Philippe Retornaz <philippe@shapescale.com>
Signed-off-by: Francois Ramu <francois.ramu@st.com>

# Conflicts:
#	drivers/gpio/gpio_stm32.h

FRASTM committed:
boards: arm: Add nucleo for nucleo_g071rb board

This patch adds support for the STM32 nucleo_g071rb board
 from STMicroelectronics.

Signed-off-by: Philippe Retornaz <philippe@shapescale.com>
Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango reviewed (changes requested):
LGTM, but please fix checkpatch first (dts_fixup.h warnings could be omitted)

FRASTM committed:
dts: remove unused clock-controllers property #17122

The 'clock-controllers' property is not specified and not used for any
generation.  Lets remove it.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: clock_control: Add STM32G0XX clock support

Add clock support for STM32G0X SoC series.

Signed-off-by: Philippe Retornaz <philippe@shapescale.com
Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
> LGTM, but please fix checkpatch first (dts_fixup.h warnings could be omitted)

done 
checkpatch — Checks passed

FRASTM commented:
> > LGTM, but please fix checkpatch first (dts_fixup.h warnings could be omitted)
> 
> done
> checkpatch — Checks passed

ydamigos commented on drivers/clock_control/clock_stm32g0.c:
```diff
@@ -0,0 +1,51 @@
+/*
```
Look very similar to `clock_stm32l4_wb.c`. Is it possible to use `clock_stm32l4_wb.c` instead of `clock_stm32g0.c`?

ydamigos replied:
@FRASTM From your comment below, it is possible to combine Kconfig.stm32l4_wb and Kconfig.stm32g0 in a  Kconfig.stm32l4_wb_g0. Could we do the same for clock_stm32l4_wb.c and clock_stm32g0.c?

ydamigos commented on drivers/clock_control/Kconfig.stm32g0:
```diff
@@ -0,0 +1,51 @@
+# Kconfig - STM32G0 PLL configuration options
```
Look very similar to `Kconfig.stm32l4_wb`. Is it possible to use `Kconfig.stm32l4_wb` instead of `Kconfig.stm32g0`?

FRASTM replied:
yes, could be with a "Kconfig.stm32l4_wb_g0"

erwango replied:
For consistency, I'd prefer we keep aligned Kconfig.stm32xx and clock_xx.c files.
So either we keep them split, either we merge both.

FRASTM committed:
drivers: flash: Add STM32G0XX flash support

Add flash support for STM32G0X SoC series.

Signed-off-by: Philippe Retornaz <philippe@shapescale.com
Signed-off-by: Francois Ramu <francois.ramu@st.com>
